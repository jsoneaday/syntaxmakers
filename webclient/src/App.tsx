import { createBrowserRouter, RouterProvider } from "react-router-dom";
import Home from "./presentation/pages/Home";
import EmpJobPosts from "./presentation/pages/employer/EmpJobPosts";
import EmpJobApplicants from "./presentation/pages/employer/EmpJobApplicants";
import DevJobSearch from "./presentation/pages/developer/DevJobSearch";
import DevAppliedJobs from "./presentation/pages/developer/DevAppliedJobs";
import DevJobFullview from "./presentation/pages/developer/DevJobFullview";
import EmpJobFullview from "./presentation/pages/employer/EmpJobFullview";
import DevProfile from "./presentation/pages/developer/DevProfile";
import EmpProfile from "./presentation/pages/employer/EmpProfile";
import ConfirmEmail from "./presentation/pages/ConfirmEmail";
import EmpJobApplicant from "./presentation/pages/employer/EmpJobApplicant";

export const DEV_ROUTE_PREFIX = "devhome";
export const EMP_ROUTE_PREFIX = "emphome";

export enum RoutePaths {
  Root = "/",
  DevJobSearch = `/${DEV_ROUTE_PREFIX}/searchjobs`,
  DevAppliedJobs = `/${DEV_ROUTE_PREFIX}/appliedjobs`,
  DevProfile = `/${DEV_ROUTE_PREFIX}/profile`,
  DevJobFullView = `/${DEV_ROUTE_PREFIX}/jobfullview`,
  EmpPostNewJob = `/${EMP_ROUTE_PREFIX}/postnewjob`,
  EmpProfile = `/${EMP_ROUTE_PREFIX}/profile`,
  EmpJobPosts = `/${EMP_ROUTE_PREFIX}/jobposts`,
  EmpJobApplicants = `/${EMP_ROUTE_PREFIX}/jobapplicants`,
  ApplicantProfile = `/${EMP_ROUTE_PREFIX}/viewdev/:id`,
  EmpJobFullView = `/${EMP_ROUTE_PREFIX}/jobfullview`,
  EmailConfirmation = "/confirm_email",
}

const router = createBrowserRouter([
  {
    path: RoutePaths.Root,
    element: <Home />,
  },
  {
    path: RoutePaths.DevJobSearch,
    element: <DevJobSearch />,
  },
  {
    path: `${RoutePaths.DevJobSearch}/:search`,
    element: <DevJobSearch />,
  },
  {
    path: RoutePaths.DevAppliedJobs,
    element: <DevAppliedJobs />,
  },
  {
    path: RoutePaths.DevProfile,
    element: <DevProfile />,
  },
  {
    path: RoutePaths.DevJobFullView,
    element: <DevJobFullview />,
  },
  {
    path: RoutePaths.EmpPostNewJob,
    element: <EmpJobFullview />,
  },
  {
    path: RoutePaths.EmpProfile,
    element: <EmpProfile />,
  },
  {
    path: RoutePaths.EmpJobPosts,
    element: <EmpJobPosts />,
  },
  {
    path: `${RoutePaths.EmpJobPosts}/:search`,
    element: <EmpJobPosts />,
  },
  {
    path: RoutePaths.EmpJobApplicants,
    element: <EmpJobApplicants />,
  },
  {
    path: RoutePaths.ApplicantProfile,
    element: <EmpJobApplicant />,
  },
  {
    path: RoutePaths.EmpJobFullView,
    element: <EmpJobFullview />,
  },
  {
    path: RoutePaths.EmailConfirmation,
    element: <ConfirmEmail />,
  },
]);

function App() {
  return <RouterProvider router={router} />;
}

export default App;
