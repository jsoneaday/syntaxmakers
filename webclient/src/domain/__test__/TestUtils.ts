import { faker } from "@faker-js/faker";

export function getFakeFullName() {
  return faker.person.fullName();
}
