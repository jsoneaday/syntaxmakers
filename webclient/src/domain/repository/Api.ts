export const API_ROOT_URL = "http://localhost:4003/v1/";
export const SALARIES_URL = `${API_ROOT_URL}salaries`;
export const APPLY_JOB_URL = `${API_ROOT_URL}apply_job`;
export const DEV_APPLIED_JOB_URL = `${API_ROOT_URL}developer_applied`;
export const LANGS_URL = `${API_ROOT_URL}languages`;
export const JOB_URL = `${API_ROOT_URL}job`;
export const JOB_UPDATE_URL = `${API_ROOT_URL}job_update`;
export const JOBS_DEV_URL = `${API_ROOT_URL}jobs_dev`;
export const JOBS_EMP_URL = `${API_ROOT_URL}jobs_emp`;
export const JOBS_SEARCH_URL = `${API_ROOT_URL}jobs_search`;
export const JOBS_APPLIED_URL = `${API_ROOT_URL}jobs_applied`;
export const INDUSTRIES_URL = `${API_ROOT_URL}industries`;
export const EMPLOYER_URL = `${API_ROOT_URL}employer`;
export const EMPLOYER_EMAIL_URL = `${API_ROOT_URL}employer_email`;
export const EMPLOYERS_URL = `${API_ROOT_URL}employers`;
export const DEVELOPER_URL = `${API_ROOT_URL}developer`;
export const DEVELOPER_EMAIL_URL = `${API_ROOT_URL}developer_email`;
export const DEVELOPERS_URL = `${API_ROOT_URL}developers`;
export const COUNTRIES_URL = `${API_ROOT_URL}countries`;
export const COMPANY_URL = `${API_ROOT_URL}company`;
export const COMPANIES_URL = `${API_ROOT_URL}companies`;
export const LOGIN_URL = `${API_ROOT_URL}login`;
export const REFRESH_TOKEN_URL = `${API_ROOT_URL}refreshtoken`;

export type OutputId = {
  id: number;
};

export type OutputBool = {
  result: boolean;
};
