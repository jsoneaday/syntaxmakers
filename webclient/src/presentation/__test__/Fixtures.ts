import DevProfile from "../models/DevProfile";
/// @ts-ignore
import { v4 as uuidv4 } from "uuid";
import { faker } from "@faker-js/faker";
import { getFakeFullName } from "../../domain/TestUtils";

export const defaultDevProfile: DevProfile = {
  key: uuidv4(),
  id: "1",
  updatedAt: new Date().toISOString(),
  userName: "ronaldo", //faker.internet.userName(),
  fullName: getFakeFullName(),
  email: faker.internet.email(),
  primaryLangId: "1",
};
