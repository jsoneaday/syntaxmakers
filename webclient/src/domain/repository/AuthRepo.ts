import { LOGIN_URL, REFRESH_TOKEN_URL } from "./Api";

export enum DevOrEmployer {
  Developer = "Developer",
  Employer = "Employer",
}

export type LoginResult = {
  message: string;
  status: number;
};

export async function login(
  devOrEmp: DevOrEmployer,
  email: string,
  password: string
) {
  try {
    const response = await fetch(LOGIN_URL, {
      method: "post",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        devOrEmp,
        email,
        password,
      }),
    });

    if (response.ok) {
      const access_token: string = await response.text();
      return {
        message: access_token,
        status: response.status,
      };
    }
    return {
      message: response.statusText,
      status: response.status,
    };
  } catch (err) {
    console.log("login failed", err);
  }

  return {
    message: "",
    status: 400,
  };
}

export async function refreshAccessToken(
  oldToken: string,
  devOrEmp: DevOrEmployer
) {
  try {
    const response = await fetch(REFRESH_TOKEN_URL, {
      method: "post",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        oldToken,
        devOrEmp,
      }),
    });

    if (response.ok) {
      const access_token: string = await response.text();
      return access_token;
    }
  } catch (err) {
    console.log("login failed", err);
  }

  return "";
}
