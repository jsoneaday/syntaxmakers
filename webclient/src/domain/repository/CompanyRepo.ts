import { COMPANIES_URL } from "./Api";

export class Company {
  constructor(
    public id: string,
    public updatedAt: string,
    public name: string
  ) {}
}

export async function getCompanies(): Promise<Company[]> {
  const response = await fetch(COMPANIES_URL, {
    method: "get",
  });

  if (response.ok) {
    return await response.json();
  }
  return [];
}
