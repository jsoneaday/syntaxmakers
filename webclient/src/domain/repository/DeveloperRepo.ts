import { DEVELOPER_URL } from "./Api";

export class Developer {
  constructor(
    public id: string, // bigint is not serializable by Redux
    public updated_at: Date,
    public user_name: string,
    public full_name: string,
    public email: string,
    public primary_lang_id: string,
    public secondary_lang_id?: string | null
  ) {}
}

export async function getDeveloper(id: string) {
  console.log(`${DEVELOPER_URL}/${id}`);
  const response = await fetch(`${DEVELOPER_URL}/${id}`, {
    method: "get",
  });

  if (response.ok) {
    const developerObj: Developer = await response.json();
    return developerObj;
  }
  return null;
}
