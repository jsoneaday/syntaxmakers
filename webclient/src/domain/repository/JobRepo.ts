import { JOBS_DEV_URL, JOBS_EMP_URL, JOB_UPDATE_URL } from "./Api";

export class Job {
  constructor(
    public id: number,
    public updatedAt: string, // comes from api as utc string
    public employerId: number,
    public employerName: string,
    public companyId: number,
    public companyName: string,
    public title: string,
    public description: string,
    public isRemote: boolean,
    public primaryLangId: number,
    public primaryLangName: string,
    public secondaryLangId: number,
    public secondaryLangName: string,
    public industryId: number,
    public industryName: string,
    public salaryId: number,
    public salary: string,
    public companyLogo?: ArrayBuffer, // normal format for file data received over wire
    public countryId?: number,
    public countryName?: string
  ) {}
}

export interface JobFormState {
  id: number;
  employerId: number;
  title: string;
  description: string;
  isRemote: boolean;
  industryId: number;
  salaryId: number;
  primaryLangId: number;
  secondaryLangId?: number;
  countryId?: number;
}

export async function getJobsByDeveloper(
  id: string,
  pageSize: number = 20,
  lastOffset: number = 0
) {
  const result = await fetch(JOBS_DEV_URL, {
    method: "post",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      id,
      pageSize,
      lastOffset,
    }),
  });

  if (result.ok) {
    const jobs: Job[] = await result.json();
    return jobs;
  }
  return [];
}

export async function getJobsByEmployer(
  id: string,
  pageSize: number = 20,
  lastOffset: number = 0
): Promise<Job[]> {
  const result = await fetch(JOBS_EMP_URL, {
    method: "post",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      id,
      pageSize,
      lastOffset,
    }),
  });

  if (result.ok) {
    return await result.json();
  }
  return [];
}

export async function updateJobPost(
  jobFormState: JobFormState,
  access_token: string
) {
  console.log("jobFormState", jobFormState);

  const result = await fetch(JOB_UPDATE_URL, {
    method: "post",
    credentials: "include",
    cache: "no-store",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${access_token}`,
    },
    body: JSON.stringify({
      id: jobFormState.id,
      employerId: jobFormState.employerId,
      title: jobFormState.title,
      description: jobFormState.description,
      isRemote: jobFormState.isRemote,
      primaryLangId: jobFormState.primaryLangId,
      industryId: jobFormState.industryId,
      salaryId: jobFormState.salaryId,
      secondaryLangId: jobFormState.secondaryLangId,
      countryId: jobFormState.countryId,
    }),
  });

  if (result.ok) {
    if (result.status === 204) {
      return "";
    }
    return await result.json();
  }
  console.log(result.statusText);
  throw new Error("Failed to update job");
}
