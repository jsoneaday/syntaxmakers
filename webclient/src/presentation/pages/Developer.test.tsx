import { faker } from "@faker-js/faker";
import { getFakeFullName } from "../../domain/__test__/TestUtils";
import { Provider as ReduxProvider } from "react-redux";
import * as DevRepo from "../../domain/repository/DeveloperRepo";
import * as JobRepo from "../../domain/repository/JobRepo";
import { Developer } from "../../domain/repository/DeveloperRepo";
import { render, waitFor, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
import configureStore from "redux-mock-store";
import { defaultDevProfile } from "../__test__/Fixtures";
import App from "../../App";

const mockStore = configureStore();
const store = mockStore({
  profile: defaultDevProfile,
});

describe("Test Developer page", () => {
  it("loads home page then switches to developer page", async () => {
    jest.spyOn(DevRepo, "getDeveloper").mockImplementation(
      () =>
        new Promise((res) => {
          res(
            new Developer(
              "1",
              new Date().toISOString(),
              faker.internet.userName(),
              getFakeFullName(),
              faker.internet.email(),
              "1"
            )
          );
        })
    );

    jest.spyOn(JobRepo, "getJobsByDevProfile").mockImplementation(
      () =>
        new Promise((res) => {
          res([
            new JobRepo.Job(
              "1",
              new Date().toISOString(),
              "1",
              getFakeFullName(),
              "1",
              faker.company.name(),
              faker.lorem.sentence(50),
              faker.lorem.sentence(100),
              true,
              "1",
              faker.animal.dog(),
              "2",
              faker.animal.bird(),
              "1",
              faker.commerce.product(),
              "1",
              "200000"
            ),
          ]);
        })
    );

    render(
      <ReduxProvider store={store}>
        <App />
      </ReduxProvider>
    );

    const devLink = screen.getByTestId("dev-link");
    await userEvent.click(devLink);

    const devPage = await waitFor(() => screen.getByTestId("developer-page"));
    expect(devPage).toBeInTheDocument();
  });
});
