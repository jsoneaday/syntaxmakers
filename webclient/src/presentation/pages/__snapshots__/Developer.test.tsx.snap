// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Developer page matches snapshot 1`] = `
<div
  class="dev-container"
  data-testid="developer-page"
>
  <nav
    class="panel-col dev-menu-container"
  >
    <span
      class="title-font left-menu-header"
    >
      @jonny
    </span>
    <a
      class="sub-title-font dev-menu-item"
    >
      <img
        class="dev-menu-icon"
        src="default"
      />
      <span>
        Saved jobs
      </span>
    </a>
    <a
      class="sub-title-font dev-menu-item"
    >
      <img
        class="dev-menu-icon"
        src="default"
      />
      <span>
        Job alerts
      </span>
    </a>
  </nav>
  <div
    class="dev-main"
  >
    <div
      class="dev-top"
    >
      <div
        class="title-font dev-header"
      >
        Developer job search
      </div>
      <div
        class="sub-title-font dev-sub-header"
      >
        Enter your preferences to find your next job
      </div>
      <div
        class="search-header"
      >
        <input
          class="search-input"
          type="text"
          value=""
        />
        <button
          class="primary-btn"
        >
          search
        </button>
      </div>
    </div>
    <div
      class="info-band"
    >
      <img
        class="dev-info-band-icon"
        src="default"
      />
      Result count 231
      <img
        class="dev-info-band-icon"
        src="default"
        style="margin-left: 1.5em;"
      />
      Date jun 16, 2023
    </div>
    <div
      class="dev-post-preview-container"
    >
      <ul>
        <ul
          class="lister-container"
        >
          <li
            class="dev-preview-item"
          >
            <div
              class="post-preview-container"
            >
              <img
                class="preview-icon"
                src=""
              />
              <div
                class="preview-content"
              >
                <div
                  class="title-font"
                >
                  Senior React Developer
                </div>
                <div
                  class="sub-title-font"
                >
                  Company A
                </div>
                <div
                  class="normal-font"
                >
                  Remote
                </div>
                <div
                  class="normal-font preview-salary"
                >
                  Base Salary: 
                  <i>
                    $200000
                  </i>
                </div>
                <div
                  class="small-font preview-timestamp"
                >
                  less than a minute ago
                </div>
              </div>
            </div>
          </li>
        </ul>
      </ul>
    </div>
  </div>
  <div
    class="panel-col job-menu-container"
  >
    <div
      class="title-font"
    >
      Promoted jobs
    </div>
    <ul>
      <ul
        class="lister-container"
      />
    </ul>
  </div>
</div>
`;
